doc/Design.md

```diff
@@ -143,3 +143,91 @@ dereferencing. Unlike shared buffers, which can only be a buffer type in a
+   the virtual alarms and virtual I2C stack, as well as drivers for
+   hardware peripherals not on the chip itself (e.g. sensors, radios,
+   etc). These drivers typically rely on the chip-specific drivers
+   through the HILs.
+
+3. System call drivers, also typically found in the `capsules`
+   crate. These are the drivers that implement a particular part of
+   the system call interfaces, and are often even more abstracted from
+   the hardware than (2) - for example, the temperature sensor system
+   call driver can use any temperature sensor, including several
+   implemented as portable peripheral drivers. We don’t have many
+   examples of this, but the system call interface is another point of
+   standardization that can be implemented in various ways. So it’s
+   perfectly reasonable to have several implementations of the same
+   system call interface that use completely different hardware
+   stacks, and therefore HILs and chip-specific drivers (e.g. a
+   console driver that operates over USB might just be implemented as
+   a different system call driver that implements the same system
+   calls, rather than trying to fit USB into the UART HIL).
+
+The connective tissue between layers are the HILs. The HIL interfaces
+are portable interfaces that are implemented in a non-portable way.
+
+The choice of particular HIL interfaces is pretty important, and we
+have some general principles we follow:
+
+1. HIL implementations get to assume this HIL is the only way the
+   device will be used. As a result, it’s generally a bad idea to have
+   several HILs that provide different interfaces to similar
+   functionality, because it will not, in general, be possible for
+   multiple drivers to use different HILs for the same device
+   simultaneously.
+
+2. HIL implementations should be fairly general. If we have an
+   interface that doesn't work very well across different hardware, we
+   probably have the wrong interface - it’s either too high level, or
+   too low level, or it’s just not flexible enough. But HILs shouldn’t
+   generally be designed to optimize for particular applications or
+   hardware, and definitely not for a particular combination of
+   applications and hardware. If there are cases where that is really
+   truly necessary, you can always implement a driver that is chip or
+   board specific, and circumvents the HILs entirely.
+
+### Split-phase operation
+
+While processes are time sliced and preemptive in Tock, the kernel is
+not. Everything is run-to-completion. That’s an important design
+choice because it allows the kernel to avoid allocating lots of stacks
+for lots of tasks, and it makes it possible to reason more simply
+about static and other shared variables.
+
+As a result, by design, all I/O operations have to be asynchronous, so
+that the kernel can be reasonably counted on to operate in a timely
+manner. We do this using TinyOS style split-phase callbacks because
```

Comment:
These aren't resolved statically; there are dynamic client registrations. I'd change this paragraph to

"Therefore, all I/O operations in the Tock kernel are asynchronous. A method call starts an operation and returns immediately. When the operation completes, the struct implementing the operation calls a callback. Tock uses callbacks rather than closures because closures typically require dynamic memory allocation, which the kernel avoids and does not generally support."